Add flag vulnerable packages spec

drewgillies opened the pull request:
Let's add this spec here in its current form so we can track changes and discussions, e.g. regarding the MVP.

drewgillies committed:
Add flag vulnerable packages spec

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+
+#### ❓ Open questions (Stage 2)
+
+* For `restore`, should it be opt-in or opt-out (preferred for security features)?
```
It should be default / opt-out. Security should be ran on every-single restore just like npm install.

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+#### ❓ Open questions (Stage 2)
+
+* For `restore`, should it be opt-in or opt-out (preferred for security features)?
+* For `dotnet add package`, define behavior (info, warn, error?) when encountering packages known to be vulnerable.
```
None. We don't do this for other audits & generally we are just advising people on restore to not use these packages & providing them a resolution. Everything should be in dotnet restore instead.

drewgillies replied:
I'll adjust the spec to reflect this, but I want to be sure that we're on the same page. `dotnet add package` includes a `dotnet restore` unless the `--no-restore` option is included (https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package). So, we will have `dotnet restore` (above) defaulting to indicating vulnerabilities, and `dotnet add package` will have the same behavior as `dotnet restore` except when `--no-restore` is used. Right?

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+
+* For `restore`, should it be opt-in or opt-out (preferred for security features)?
+* For `dotnet add package`, define behavior (info, warn, error?) when encountering packages known to be vulnerable.
+* Need for another command (e.g. `dotnet nuget audit`) to create an audit report for deprecated, vulnerable and may be outdated packages. And a way to fix the vulnerabilities with a single command/click.
```
Proposed `dotnet nuget audit` & `dotnet nuget audit fix` for resolution.

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+| Feature | State |
+|:--- |:--- |
+| A way to specify another source for vulnerability information | Backlog |
+| `why` command to understand why a package is shown as a dependency for my project | Backlog |
```
This is a separate feature entirely.

drewgillies replied:
I'll remove this.

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+
+| Feature | State |
+|:--- |:--- |
+| A way to specify another source for vulnerability information | Backlog |
```
I don't believe we have any other sources in mind today & likely a couple years from another standard. I don't think we need this on a backlog unless we plan to commit to it relatively soon.

drewgillies replied:
I'll remove this too :)

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+|:--- |:--- |
+| A way to specify another source for vulnerability information | Backlog |
+| `why` command to understand why a package is shown as a dependency for my project | Backlog |
+| Ability to report a vulnerability on nuget.org | Discuss |
```
NuGet package page requires a "Report a vulnerability" feature that is triaged by someone on the NuGet team.

Otherwise we need to provide instructions for https://docs.github.com/en/github/managing-security-vulnerabilities/creating-a-security-advisory

drewgillies replied:
I think the second option is more robust as it gives a single point of truth. In a scenario where we act as a moderator, there's more opportunity for lag before a new vulnerability is surfaced in our clients.

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+| A way to specify another source for vulnerability information | Backlog |
+| `why` command to understand why a package is shown as a dependency for my project | Backlog |
+| Ability to report a vulnerability on nuget.org | Discuss |
+| An API that lists all vulnerable nuget.org packages with the vulnerability information | Discuss |
```
This is included in my design doc, maybe we can weave it in here.

drewgillies replied:
I've replaced this open question with the relevant excerpt from your doc.

JonDouglas commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,142 @@
+| `why` command to understand why a package is shown as a dependency for my project | Backlog |
+| Ability to report a vulnerability on nuget.org | Discuss |
+| An API that lists all vulnerable nuget.org packages with the vulnerability information | Discuss |
+| Show vulnerability information as part of search results | Discuss |
```
If anything, it should show as an icon similar to the deprecated package in the package details.

nkolev92 replied:
automatic reporting during restore will have a cost. 
Whether we're willing to pay for that cost is important to discuss.

drewgillies committed:
Update doc to reflect PM feedback

drewgillies commented:
@JonDouglas I'm keen on landing the MVP here, and there's one piece of it that we need to address: transitive dependencies in the PM UI for `PackageReference` projects. My concern is that we are wanting to introduce vulnerability functionality into a scenario that doesn't exist yet in any form (i.e. there is currently no `PackageReference` transitive dependency story in PM UI for us to extend). 

To go back a step, we have these discussions around how we want to see these dependencies surfaced in PM UI, and whether there should be more user control on dependency resolution (along with whether there should be promotion-to-top-level and a related prune command):
* https://github.com/NuGet/Home/issues/5887
* https://github.com/NuGet/Home/issues/5553

I propose that we remove all PM UI transitive dependency vulnerability work from the MVP until PM UI transitive dependencies are in-product for this work to extend. 

If we wish instead to make general PM UI transitive dependency work a pre-requisite for the PM UI vulnerability work, I think it would be more helpful to describe it as such and spec it separately. Then we could do PM UI vulnerability work as a stage 2 piece (i.e. not MVP). I feel we run the risk of diluting the vulnerability work here.

/cc @aortiz-msft @skofman1

JonDouglas commented:
@drewgillies I agree with you and also brought this up in various email threads. I believe this spec is simply too large for a MVP considering the VS PM UI needing to support an entirely new UI concept that hasn't been specced by itself outside of various feature specs that reference it. That to me is scope creep at it's finest.

I do not believe the right approach is to simply lump this feature into existing feature(s), as it requires involvement by PM, UX, and Engineering to bring into Visual Studio, and then for existing features to leverage it to their needs at a later stage.

Given vulnerability is not the only spec that references this work, we should consider this a piece of work that these specs "depend on" & not necessarily part of the spec itself.

drewgillies committed:
Removed PM UI transitive dependency vulnerability work from the MVP

anangaur commented:
Irrespective of how we do it, IMO the vulnerability feature Phase 1 will be incomplete unless we are able to show all vulnerabilities in the full package graph for a project/solution in Visual Studio.

JonDouglas commented:
@anangaur Referring back to the original spec:

✔️ In scope (Stage 1)
> Show transitive packages for PackageReference type projects in PMUI

❌ Out of scope (Stage 1)
> Display vulnerability information for transitive packages in PMUI (additional UX complexity)

There does not exist a spec/design of the "in scope" work for showing transitive packages in PMUI. This is troublesome for many reasons because it affects the core Installed & Update experiences in PMUI. The [screenshots on the original spec](https://github.com/NuGet/Home/wiki/Flag-vulnerable-packages#display-vulnerabilities-in-visual-studio-pmui) I do not believe are sufficient enough for the larger body of work of showing transitive packages in PMUI.

For example:

- Can a user interact with transitive packages like they do top-level packages?
- Can a user see how it was implicitly referenced?
- What about selecting a transitive package? Do we show package details or not?
- Can a user install the implicit package explicitly to "promote" to a top-level package?
- etc

We should consider https://github.com/NuGet/Home/issues/5887 & https://developercommunity.visualstudio.com/content/idea/538119/add-dependency-tree-view-for-packagereferences.html as requirements to fully complete this feature work. This work I believe needs it's own design & spec itself given the body of work.

nkolev92 commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,147 @@
+#### ✔️ In scope (Stage 2)
+
+* For CLI:
+  * For `dotnet restore`, vulnerability reporting is opt-out. By default vulnerabilities are reported during a restore operation.
```
I think this was brought up when the original design and I don't want to block/try to answer it here, but given the registration approach of reporting, reporting vulnerabilities at restore will incur a performance cost. 

We'll need a detailed design when we do get to phase 2. Personally I think we sohuld call it out the same way we're calling out the transitive dependency UX challenges.

JonDouglas replied:
Agreed. I'll make a note here given it is our eventual goal at what cost though? :)

JonDouglas commented:
@drewgillies I will do my best to see if there's an existing spec anywhere. I've looked and I don't think there is one. I may have to write one & include it as another proposal. I don't believe we should treat this as blocking however, we can always just show Top-Level dependencies vulnerabilities in the UI to start & add vulnerability information for transitive dependencies in Stage 2.

> Spec the PM UI transitive dependencies work in general. If we can get that spec completed and signed off (i.e. added here to home/designs), then we have a starting point. If we do indeed have a firm spec, let's get it on the table in a separate PR here. If we don't, this should be our first step.

aortiz-msft commented on designs/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,147 @@
+* Display vulnerability information for *transitive packages* in PMUI (additional UX complexity)
+* Expose vulnerability information *publicly* on nuget.org's package details page and version history (disclosure policies TBD)
+
+### Stage 2
```
It would be good if we could move Stage 2 to another design spec that we can review when we are ready to take on that work. As part of this spec, we should only mention potential enhancements in Stage 2 that have an impact on the initial design.

JonDouglas reviewed (approved):
Approving for Stage 1 work. Stage 2 needs a separate spec.

anangaur commented:
We should also update this [announcement ](https://github.com/NuGet/Announcements/issues/39)or put another new one?

drewgillies commented:
@JonDouglas if stage 2 needs a separate spec, we should probably pull it out of this one. Agreed?

aortiz-msft commented:
> @JonDouglas if stage 2 needs a separate spec, we should probably pull it out of this one. Agreed?

Yes please :)

drewgillies committed:
Remove stage 2

drewgillies commented:
@JonDouglas @aortiz-msft Done. I've moved the vulnerable package transitive dependencies work into the PM UI for stage 1 since that seems to be becoming the thinking.

aortiz-msft commented:
@drewgillies - anything needed here?

aortiz-msft commented:
@drewgillies - Is there work left for this issue?

JonDouglas commented:
Hello!

I'm working to cleanup stale PRs on `NuGet/Home`. 

Actions to take:

- If this PR **should be merged** into `NuGet/Home`, please assign a handful of reviewers or ask for another review to approve a merge.
- If this PR **should be abandoned**, please close it.
- If this PR **should remain open**, please bump the conversation forward or leave a comment for this to not be considered abandoned in a future pass.

If there's anything I can do to help out, please let me know!

drewgillies commented:
@aortiz-msft -- apologies for missed messages on this. I'll review personally shortly to see if any services/nuget.org/dotnet CLI pieces need adjusting, and then I'll submit this for an additional review cycle per @JonDouglas's recommendation.

drewgillies committed:
Removed out of scope item on public display of vulnerabilities, corrected link

drewgillies commented:
I've cleaned this up a little (removed non-goal of publishing vulnerabilities publicly on nuget.org, corrected a spec link) and moved this spec to the proposed directory. Re-requesting for final review per above. 
Easy link for viewing updated spec: https://github.com/NuGet/Home/blob/dev-drewgil-add-flagvulnerablepackages-spec/proposed/2020/PackageVulnerability/FlagVulnerablePackages.md
FYI @JonDouglas @aortiz-msft

nkolev92 commented on proposed/2020/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,99 @@
+
+Some enhancements to this feature are out of scope for the initial implementation of this feature because there are several unknowns still to be clarified. See also the *Open Questions* section below.
+
+* Enable `restore` to Inform or Warn (see Stage 2)
```
nit: Stage 2 is not defined, but referred to :)

I imagine you just want to say it'll be considered at a future time.

drewgillies replied:
Thanks, @nkolev92 --this is removed now.

anangaur commented on proposed/2020/PackageVulnerability/FlagVulnerablePackages.md:
```diff
@@ -0,0 +1,99 @@
+  * **Automatically ingest and refresh NuGet package vulnerability data from GitHub**
+  * **Surface package vulnerability information on-demand**:
+    * as part of package metadata in `v3` registration blobs (`JSON`)
+    * by showing vulnerability information to package owners (only!) on package details page and version history
```
Are we still planning to show this info to authors only? /cc: @JonDouglas

drewgillies replied:
Good catch, @anangaur --I've removed this now, as well as the email bullet.

drewgillies committed:
Addressed feedback

drewgillies commented:
If there are no further concerns about this, I'll go ahead and merge it. @aortiz-msft @nkolev92 @anangaur @JonDouglas

drewgillies commented:
Thanks everyone for your contributions! Merging.
